### PR TITLE
target/ppc: Vector Extract and Vector Insert

### PR DESCRIPTION
**Pull request by mferst**

### CODE REVIEW

**Review by alqotel — APPROVED** [review]
LGTM

**Review by lffpires — COMMENTED** [review]
Partial review (for now, just for "target/ppc: Implement Vector Extract Double to VSR using GPR index insns")

File: target/ppc/translate/vmx-impl.c.inc [hunk 1]
```diff
@@ -1238,6 +1238,41 @@ GEN_VXFORM_DUAL(vspltish, PPC_ALTIVEC, PPC_NONE,
...
+static bool do_vextdx(DisasContext *ctx, arg_VA *a, int size, bool right,
+                      void (*gen_helper)(TCGv_ptr, TCGv_ptr, TCGv_ptr, TCGv))
+{
+    TCGv rc;
+    TCGLabel *l0;
+
+    REQUIRE_INSNS_FLAGS2(ctx, ISA310);
+    REQUIRE_VECTOR(ctx);
+
+    rc = tcg_temp_local_new();
```
[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/translate/vmx-impl.c.inc:1394]: AFAIK, we don't have something like `gen_invalid` to call from the helper, so we'd need to return something to indicate the result and branch accordingly. It would spare us the local temp, but we'd have an expensive helper call for something that can be solved in TCG.

[the remark this answers is not in the document]
Comment (lffpires) [reply target/ppc/translate/vmx-impl.c.inc:1394]: gen_invalid? The branch is just skipping the result calculation when idx is too large, isn't it?

File: target/ppc/int_helper.c [hunk 2]
```diff
@@ -1528,6 +1528,59 @@ VEXTU_X_DO(vextuhrx, 16, 0)
...
+        int len_a = array_size - ofs, len_b = sz - len_a;                   \
+                                                                            \
+        memmove(&tmp.u8[array_size / 2], &b->u8[array_size - len_b], len_b);\
+        memmove(&tmp.u8[array_size / 2 + len_b], &a->u8[0], len_a);         \
+    }                                                                       \
+                                                                            \
+    *t = tmp;                                                               \
+}
+#endif
+VEXTDVLX(vextdubvlx, 1)
```
[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/int_helper.c]: I tried to apply some of this in https://github.com/PPC64/qemu/pull/25/commits/228f6aa37ea8d9afd7e891345ae5b769c2443ebf
- Checks within the helper, and also log a guest error for invalid indexes
- I prefer to keep all cases in the "main" macro, handling the non-crossing case in a ternary operator seems less readable
- It seems easier to pass the type and use `sizeof ` instead of the size in bits

File: target/ppc/int_helper.c [hunk 3]
```diff
@@ -1529,56 +1529,49 @@ VEXTU_X_DO(vextuwrx, 32, 0)
...
-VEXTDVLX(UB, 1)
-VEXTDVLX(UH, 2)
-VEXTDVLX(UW, 4)
-VEXTDVLX(D, 8)
+#define VEXTDVLX(NAME, TYPE) \
+void glue(glue(helper_VEXTD, NAME), VLX)(CPUPPCState *env, ppc_avr_t *t,       \
+                                         ppc_avr_t *a, ppc_avr_t *b,           \
+                                         target_ulong index)                   \
+{                                                                              \
+    int array_size = ARRAY_SIZE(t->u8), elem_size = sizeof(TYPE);              \
```
Comment (lffpires) [inline target/ppc/int_helper.c:1541]: Probably doesn't make any real difference (except maybe for people reading the code), but I would make these 'const int'.

File: target/ppc/insn32.decode [hunk 4]
```diff
@@ -37,6 +37,9 @@
 &VX             vrt vra vrb
 @VX             ...... vrt:5 vra:5 vrb:5 .......... .   &VX
 
+&VX_uim4        vrt uim vrb
+@VX_uim4        ...... vrt:5 - uim:4 vrb:5 ...........  &VX_uim4
```
Comment (lffpires) [inline target/ppc/insn32.decode:41]: We should keep just the field definitions here, and put the '-'s (don't care) in the instruction definitions. In this case, they're already there, so just need to replace the single '-' with '.'

File: target/ppc/translate/vmx-impl.c.inc [hunk 5]
```diff
@@ -1231,12 +1227,62 @@ GEN_VXFORM_DUAL(vsplth, PPC_ALTIVEC, PPC_NONE,
...
 GEN_VXFORM_DUAL(vspltw, PPC_ALTIVEC, PPC_NONE,
                 vextractuw, PPC_NONE, PPC2_ISA300);
-GEN_VXFORM_DUAL(vspltisb, PPC_ALTIVEC, PPC_NONE,
-                vinsertb, PPC_NONE, PPC2_ISA300);
-GEN_VXFORM_DUAL(vspltish, PPC_ALTIVEC, PPC_NONE,
-                vinserth, PPC_NONE, PPC2_ISA300);
-GEN_VXFORM_DUAL(vspltisw, PPC_ALTIVEC, PPC_NONE,
-                vinsertw, PPC_NONE, PPC2_ISA300);
+
+static bool trans_VINSERTB(DisasContext *ctx, arg_VX_uim4 *a)
```
Comment (lffpires) [inline target/ppc/translate/vmx-impl.c.inc:1231]: Maybe use a macro to create trans_VINSERT{B,H,W,D}?

File: target/ppc/translate/vmx-impl.c.inc [hunk 6]
```diff
@@ -1284,6 +1284,39 @@ static bool trans_VINSERTD(DisasContext *ctx, arg_VX_uim4 *a)
...
+static bool do_vinsvx(DisasContext *ctx, arg_VX *a, int size, bool right,
+                      void (*gen_helper)(TCGv_ptr, TCGv_ptr, TCGv_i32))
+{
+    TCGv_i32 index;
+    TCGLabel *l0;
+
+    REQUIRE_INSNS_FLAGS2(ctx, ISA310);
+    REQUIRE_VSX(ctx);
+
+    index = tcg_temp_local_new_i32();
```
Comment (lffpires) [inline target/ppc/translate/vmx-impl.c.inc:1302]: Similar to what was done for vextdu*, can we get rid of the branch and check ``index`` inside the helper?

File: target/ppc/int_helper.c [hunk 7]
```diff
@@ -1645,23 +1645,23 @@ void helper_vslo(ppc_avr_t *r, ppc_avr_t *a, ppc_avr_t *b)
 
 #if defined(HOST_WORDS_BIGENDIAN)
 #define VINSERT(suffix, element)                                            \
-    void helper_vinsert##suffix(ppc_avr_t *r, ppc_avr_t *b, uint32_t index) \
+    void helper_VINSERT##suffix(ppc_avr_t *r, ppc_avr_t *b, uint32_t index) \
     {                                                                       \
         memmove(&r->u8[index], &b->u8[8 - sizeof(r->element[0])],           \
```
Comment (lffpires) [inline target/ppc/int_helper.c:1666]: This is not directly related to your change, but maybe a good opportunity to use assignments instead of memmove, similar to what was done for vextdu*?

File: target/ppc/int_helper.c [hunk 8]
```diff
@@ -1643,6 +1643,29 @@ void helper_vslo(ppc_avr_t *r, ppc_avr_t *a, ppc_avr_t *b)
 #endif
 }
 
+#if defined(HOST_WORDS_BIGENDIAN)
+#define VINSX(suffix, type)                                                 \
+    void glue(glue(helper_vins, suffix), x)(ppc_avr_t *r,  target_ulong b,  \
+                                            uint32_t index)                 \
+    {                                                                       \
+        type src = b;                                                       \
+        memmove(&r->u8[index], &src, sizeof(type));                         \
```
Comment (lffpires) [inline target/ppc/int_helper.c:1645]: Use assignments instead of memmove, similar to what was done for vextdu*?

File: target/ppc/translate/vmx-impl.c.inc [hunk 9]
```diff
@@ -1317,6 +1317,41 @@ TRANS(VINSBVRX, do_vinsvx, 1, true, gen_helper_VINSERTB)
...
+static bool do_vinsx(DisasContext *ctx, arg_VX *a, int size, bool right,
+                     void (*gen_helper)(TCGv_ptr, TCGv, TCGv_i32))
+{
+    TCGv_i32 index;
+    TCGLabel *l0;
+
+    REQUIRE_INSNS_FLAGS2(ctx, ISA310);
+    REQUIRE_VSX(ctx);
+
+    index = tcg_temp_local_new_i32();
```
Comment (lffpires) [inline target/ppc/translate/vmx-impl.c.inc:1335]: And get rid of the branch?

File: target/ppc/translate/vmx-impl.c.inc [hunk 10]
```diff
@@ -1352,6 +1352,36 @@ TRANS(VINSHRX, do_vinsx, 2, true, gen_helper_VINSHX)
 TRANS(VINSWRX, do_vinsx, 4, true, gen_helper_VINSWX)
 TRANS(VINSDRX, do_vinsx, 8, true, gen_helper_VINSDX)
 
+static bool trans_VINSW(DisasContext *ctx, arg_VX_uim4 *a)
```
Comment (lffpires) [inline target/ppc/translate/vmx-impl.c.inc:1361]: Only 2 functions, so no big deal. But we could also use a macro here to generate trans_VINS{W,D}

**Comment by mferst** [comment]
Replaced by #30
